pkg/prove/proof.go

```diff
@@ -203,3 +208,96 @@ func splitIntoRows(squareSize uint64, s []shares.Share) [][]shares.Share {
 	}
 	return rows
 }
+
+// SharesInclusion generates an nmt inclusion proof for a set of shares to the data root.
+// expects the shares range to be pre-validated.
+// Note: only supports inclusion proofs for shares belonging to the same namespace.
+func SharesInclusion(
+	codec rsmt2d.Codec,
+	allRawShares []shares.Share,
+	squareSize uint64,
+	namespaceID namespace.ID,
+	startShare uint64,
+	endShare uint64,
+) (types.SharesProof, error) {
+	startRow := startShare / squareSize
+	endRow := endShare / squareSize
+	startLeaf := startShare % squareSize
+	endLeaf := endShare % squareSize
+
+	eds, err := da.ExtendShares(squareSize, shares.ToBytes(allRawShares))
+	if err != nil {
+		return types.SharesProof{}, err
+	}
+
+	// create the binary merkle inclusion proof, for all the square rows, to the data root
+	rootHash, allProofs := merkle.ProofsFromByteSlices(append(eds.RowRoots(), eds.ColRoots()...))
+	rowsProofs := make([]*merkle.Proof, endRow-startRow+1)
+	rowsRoots := make([]tmbytes.HexBytes, endRow-startRow+1)
+	for i := startRow; i <= endRow; i++ {
+		rowsProofs[i-startRow] = allProofs[i]
+		rowsRoots[i-startRow] = eds.RowRoots()[i]
```

Comment:
We probably don't want to call `eds.RowRoots()` multiple times, since it does a deep copy of underlying array https://github.com/celestiaorg/rsmt2d/blob/master/extendeddatasquare.go#L183

Reply:
[3eb87ca](https://github.com/celestiaorg/celestia-app/pull/1042/commits/3eb87ca4d9368fbcdcbc14187f3022ba5d5577ed)